### ISSUE TITLE
"No role found in SamlAssertion" when logging in via Okta

### ISSUE THREAD

**Issue by kylemcleland-fanduel** [aws/amazon-redshift-python-driver#55]
# Resolution

I've resolved this issue on my own local version. 
Solution: in the file: /redshift_connector/plugin/saml_credentials_provider.py

Line 99: 
>arns = value.contents[0].split(", ")
to
>arns = value.contents[0].split(",  ")

This worked in my case because `value.contents[0]` was a list with a trailing space after the comma. If we also want to be safe capture cases where there is not a trailing space, we can use another solution

## Driver version
Python 3.8

## Redshift version
PostgreSQL 8.0.2 on i686-pc-linux-gnu, compiled by GCC gcc (GCC) 3.4.2 20041017 (Red Hat 3.4.2-6.fc3), Redshift 1.0.29551

## Client Operating System
MacOS

## Python version
Python 3.8

## Table schema
n/a

## Problem description
Running the following code:

```
import redshift_connector

gdp_conn = redshift_connector.connect(
    iam=True,
    database="mydb",
    cluster_identifier="mycluster",
    credentials_provider='OktaCredentialsProvider',
    user="myuser@domain.com",
    password="mypassword",
    idp_host="myidp",
    app_id="someverylong/string,
    region="us-east-1",
    role_arn="arn:aws:iam::{myaccountid}:role/{my_role_to_assume}",
)
```

1. Expected behaviour: A connection is returned

2. Actual behaviour: An error is returned. Looking at the debug log, it looks like it is succesfully authenticating through okta, but failing to find a proper role. It looks like it is getting to `get_credentials` within `saml_credentials_provider.py`. It's calling `self.refresh()` on line 62
```
    def get_credentials(self: "SamlCredentialsProvider") -> CredentialsHolder:
        key: str = self.get_cache_key()
        if key not in self.cache or self.cache[key].is_expired():
            try:
                self.refresh()
            except Exception as e:
                _logger.error("Refreshing IdP credentials failed: {}".format(str(e)))
                raise InterfaceError(e)
```
Since there's no key in self.cache (or it's expired), it's calling `refresh()`, but apparently refresh() is failing to return any role, so when self.refresh() gets to this code block:

```
        roles: typing.Dict[str, str] = {}
        for attr in attrs:
            name: str = attr.attrs["Name"]
            values: typing.Any = attr.findAll("AttributeValue")
            if name == "https://aws.amazon.com/SAML/Attributes/Role":
                for value in values:
                    arns = value.contents[0].split(",")
                    role: str = ""
                    provider: str = ""
                    for arn in arns:
                        if role_pattern.match(arn):
                            role = arn
                        if provider_pattern.match(arn):
                            provider = arn
                    if role != "" and provider != "":
                        roles[role] = provider

        if len(roles) == 0:
            raise InterfaceError("No role found in SamlAssertion")
```
it is not adding any role to the roles dict, therefor len(roles) == 0 and an error is raised.

3. Error message/stack trace:
```
Traceback (most recent call last):
  File "redshift_test.py", line 6, in <module>
    gdp_conn = redshift_connector.connect(
  File "/Users/kylemcleland/.pyenv/versions/3.8.10/lib/python3.8/site-packages/redshift_connector/__init__.py", line 326, in connect
    IamHelper.set_iam_properties(info)
  File "/Users/kylemcleland/.pyenv/versions/3.8.10/lib/python3.8/site-packages/redshift_connector/iam_helper.py", line 196, in set_iam_properties
    IamHelper.set_iam_credentials(info)
  File "/Users/kylemcleland/.pyenv/versions/3.8.10/lib/python3.8/site-packages/redshift_connector/iam_helper.py", line 283, in set_iam_credentials
    credentials: CredentialsHolder = provider.get_credentials()
  File "/Users/kylemcleland/.pyenv/versions/3.8.10/lib/python3.8/site-packages/redshift_connector/plugin/saml_credentials_provider.py", line 63, in get_credentials
    raise InterfaceError(e)
redshift_connector.error.InterfaceError: No role found in SamlAssertion
```

4. Any other details that can be helpful:

## Reproduction code
In part 3. above

**Comment by tomjansto** [comment]
Sir,
if you change the regex for both the role and provider in the refresh() method of the saml_credentials_provider.py file from

        role_pattern = re.compile(r"arn:aws:iam::\d*:role/\S+")
        provider_pattern = re.compile(r"arn:aws:iam::\d*:saml-provider/\S+")

to this:

        role_pattern = re.compile(r"arn:aws.*:iam::\d*:role/\S+")
        provider_pattern = re.compile(r"arn:aws.*:iam::\d*:saml-provider/\S+")

do you see that you are getting a role and provider from the SAML packet?  we are using an ADFS integration solution, and the SAML packet had an arn that was not aws, but rather aws-xxxxxxxxxx, so the regex had to be adjusted in order to get the role and provider properly extracted from the SAML packet.

note, i am by no means a git-commit worthy python developer, i'm whacking away trying to get that redshift-connnector drive functioning for an ADFS integrated IAM/IDP provider scenario.  ​i just saw this and had run into a similar situation.

tom

**Comment by kylemcleland-fanduel** [comment]
I do think it is an issue with the regex in the line you have provided. I tested your regex as well, but that change did not seem necessary. It looks 

For reference, here is the values going into the loop `for arn in arns:`

arns = ['arn:aws:iam::999999999999:saml-provider/provider_role_redacted', ' arn:aws:iam::990163193419:role/Role_Redacted']

so here is the relevant portion of `def refresh()` within `saml_credentials_provider`


```
role_pattern = re.compile(r"arn:aws:iam::\d*:role/\S+")
provider_pattern = re.compile(r"arn:aws:iam::\d*:saml-provider/\S+")
...
for arn in arns:
    if role_pattern.match(arn):
        role = arn
    if provider_pattern.match(arn):
        provider = arn
...
print("role", role)
print("provider", provider)

>>role 
>>provider arn:aws:iam::990163193419:saml-provider/okta_test
```



so role is failing the regex check and is failing to declare a variable. I'm trying to troubleshoot why that is, because using a regex checker, it looks to match the first item in my arns list.

**Comment by tomjansto** [comment]
is the leading space on the role arn you have in there causing a regex failure by any chance?

**Comment by kylemcleland-fanduel** [comment]
@tomjansto - yup! That's exactly the issue. 

in the same function, it is doing 

`arns = value.contents[0].split(",")`

But value.contents[0] = "arn:aws:iam::999999999999:saml-provider/provider_role_redacted,  arn:aws:iam::999999999999:role/Role_Redacted"

So the .split(",") is keeping the trailing whitespace.
Changing to `.split(", ")` fixes that issue. 


Cheers for that!  Unfortunately after fixing this issue, I'm now running into another issue on the same `refresh` function. 

```
credentials.set_metadata(self.read_metadata(doc))
```

Is now failing. Trying to troubleshoot that portion now.

**Comment by Brooke-white** [comment]
Hey @kylemcleland-fanduel and @tomjansto ,

Thank you both for looking into this issue! I will pass the note about the regex for SAML role and provider along to the driver team and see what we can do here in terms of its flexibility.

**Comment by Brooke-white** [comment]
this fix has been included in v2.0.888, thanks Kyle!
